### ISSUE TITLE
Avoid loading hidden images

### ISSUE THREAD

**Issue by abraham** [tjoskar/ng-lazyload-image#194]
I'm looking for a way to avoid loading images on elements that are `display: none;`.

You can see an example of the issue on https://slides.today/.

1 Load https://slides.today/
2 Open dev console
3 Select "angular" tag

The expectation is that a single additional image will be loaded since there is a single Angular card visible. However you will see about a dozen images actually loaded from cards that are currently hidden.

I tried messing around with `isVisible` to check the computed styles but that was only getting called once on first scroll event. So as the visibility on cards changed `isVisible` would never calculate again if they had changed.

**Comment by tjoskar** [comment]
Hi @abraham,

One of the features of `ng-lazyload-image` is to load the image before it is shown. In order to do that, I create an [`Img` instance](https://github.com/tjoskar/ng-lazyload-image/blob/9a07a3423fdba7151c8344b51459e873d56f871c/src/lazyload-image.ts#L36-L44) and load the image without knowing if the image is visible or not (I just know that it is in the viewport).  

We could make sure that the images are only loaded if they are visible (and not only in the viewport) but since they are no way of knowing if the image changes the visibility we can't know if the image should be loaded later on (without polling). 

However, since angular are using zone.js it should be aware of all changes and you should be able to insert the element into the DOM when it becomes visible:
```html
<img [defaultImage]="defaultImage" [lazyLoad]="lazyImage" #image *ngIf="image.offsetParent !== null">
```

But I guess you are looping over your images which will make it a bit more tricky.

**Comment by abraham** [comment]
I've got some elements that are expensive to render so I render them all once. When a user selects a tag each element will add a class that sets `display: none` if it is not related to the selected tag.

I played around a little more and found that adding this to [`isVisible`](https://github.com/tjoskar/ng-lazyload-image/blob/9a07a3423fdba7151c8344b51459e873d56f871c/src/lazyload-image.ts#L11) gets the behaviour I'm looking for. When an element is `display: none` it seems to always get `0,0,0,0` for it's bounding rect.

```
    if (top === 0 && bottom === 0 && left === 0 && right === 0) {
      return false;
    }

```

If this is a change you're interested in I can open a PR.

**Comment by tjoskar** [comment]
Yeah, we can check if the image is actually visible before we try to loads it but we have no idea when the image becomes visible, later on.

Let's take this use case:
When the user visits your site there are two images. Both are in the viewport but only one is visible (the other is `display: none`). `ng-lazyload-image` will kick in on both images but will only load the visible image (if we check it is visible, according to your change above). So far, so good.
The user now clicks on a tab that toggles the visibility of the images, so the one that wasn't visible becomes visible. `ng-lazyload-image` will however not be notified about the visibility change so it will not be loaded until the user starts scrolling (it possible). 

This could, however, be solved by notifying `ng-lazyload-image` when the user clicks on the tab. With something like:

```js
// Controller

const click$ = new Subject();
const scrollAndClick$ = Observable.merge(
  Observable.fromEvent(window, 'scroll')
  click$
);

// when the user change tab:
click$.next();
```

```html
<img [lazyLoad]="lazyLoadImage" [scrollObservable]="scrollAndClick$" />
```

How does that sound?

**Comment by dawidskowron** [comment]
I had the same problem.
This solution works for me, thanks.
